File: WEEK 1/Problem Set/Credit/credit.c
#include <cs50.h>
#include <stdio.h>

int check_card(long);
void check_flag(long);
int count_numbers(long);
int first_number(long);
int second_number(long);


int main(void)
{
    long card_num = get_long("Insert your card number: ");

    check_card(card_num);
    if (check_card(card_num) == 1) // number 1 represents true
    {
        check_flag(card_num);
    }
    else
    {
        printf("INVALID\n");
    }
}

int check_card(long card)
{
    // declaration of variables
    int sum = 0, sum2 = 0;
    int curr_digit = 0;
    int curr_sum = 0;

    while (card > 0)
    {
        curr_digit = card % 10; // first sum
        sum2 += curr_digit;

        card /= 10; // second sum
        curr_digit = card % 10;
        curr_sum = curr_digit * 2;
        if (curr_sum > 9)
        {
            sum += (curr_sum % 10) + (curr_sum / 10); // sum +=  last digit + first digit of the current digit
        }
        else
        {
            sum += curr_sum;
        }

        card /= 10; // next number for next loop
    }
    int sumt = sum + sum2;

    if (sumt % 10 == 0) // check if the last number of the total sum iguals to zero and return bool true of false
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

int count_numbers(long card)
{
    int count = 0;
    while (card > 0)
    {
        card = card / 10;
        count++;
    }

    return count;
}

int second_number(long card)
{
    while (card >= 100)
    {
        card = card / 10;
    }
    card %= 10;

    return card;

}

int first_number(long card)
{
    while (card >= 10)
    {
        card = card / 10;
    }

    return card;
}

void check_flag(long card)
{
    if (count_numbers(card) == 16)
    {
        if (first_number(card) == 4)
        {
            printf("VISA\n");
        }
        else if (first_number(card) == 5)
        {
            if (second_number(card) >= 1 && second_number(card) <= 5)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else if (count_numbers(card) == 15)
    {
        if (first_number(card) == 3)
        {
            if (second_number(card) == 4 || second_number(card) == 7)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else if (count_numbers(card) == 13)
    {
        if (first_number(card) == 4)
        {
            printf("VISA\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: WEEK 6/Problem Set/sentimental-credit/credit.py
card_str = input("Card Number: ")
lenght_card = len(card_str)
if lenght_card not in [13, 15, 16] or not card_str.isdigit():
    print("INVALID")
    exit(1)

i = lenght_card - 1
sum = 0
sum2 = 0

for index, digit in enumerate(card_str[::-1]):
    if index % 2 == 0:
        sum2 += int(digit)
    else:
        curr_digit = int(digit) * 2
        if curr_digit > 9:
            sum += (curr_digit % 10) + (curr_digit // 10)
        else:
            sum += curr_digit

if (sum + sum2) % 10 == 0:
    if card_str[0:2] in ["51", "52", "53", "54", "55"] and lenght_card == 16:
        print("MASTERCARD")
    elif card_str[0:2] in ["34", "37"] and lenght_card == 15:
        print("AMEX")
    elif card_str[0] == "4" and lenght_card in [13, 16]:
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")
